hypothesis-python/src/hypothesis/searchstrategy/featureflags.py

```diff
@@ -0,0 +1,118 @@
+        # no features are enabled, which are vanishingly unlikely in the
+        # original model.
+        #
+        # We implement this as a single 8-bit integer and enable features which
+        # score >= that value. In particular when self.__baseline is 0, all
+        # features will be enabled. This is so that we shrink in the direction
+        # of more features being enabled.
+        self.__baseline = data.draw_bits(8)
+
+    def is_enabled(self, name):
+        """Tests whether the feature named ``name`` should be enabled on this
+        test run."""
+        if self.__data.frozen:
+            # Feature set objects might hang around after data generation has
+            # finished. If this happens then we just report all new features as
+            # enabled, because that's our shrinking direction and they have no
+            # impact on data generation if they weren't used while it was
+            # running.
+            try:
+                return self.__is_value_enabled(self.__decisions[name])
+            except KeyError:
+                return True
+
+        data = self.__data
+
+        data.start_example(label=FEATURE_LABEL)
+        if name in self.__decisions:
+            # If we've already decided on this feature then we don't actually
+            # need to draw anything, but we do write the same decision to the
+            # input stream. This allows us to lazily decide whether a feature
+            # is enabled, because it means that if we happen to delete the part
+            # of the test case where we originally decided, the next point at
+            # which we make this decision just makes the decision it previously
+            # made.
+            value = self.__decisions[name]
+            data.draw_bits(8, forced=value)
+        else:
+            # If the baseline is 0 then everything is enabled so it doesn't
+            # matter what we have here and we might as well make the shrinker's
+            # life easier by forcing it to zero.
+            if self.__baseline == 0:
+                value = 0
+                data.draw_bits(8, forced=0)
+            else:
+                value = data.draw_bits(8)
+            self.__decisions[name] = value
+        data.stop_example()
+        return self.__is_value_enabled(value)
+
+    def __is_value_enabled(self, value):
+        """Check if a given value drawn for a feature counts as enabled. Note
+        that low values are more likely to be enabled. This is again in aid of
+        shrinking open. In particular a value of 255 is always enabled."""
+        return (255 - value) >= self.__baseline
+
+    def __repr__(self):
+        enabled = []
+        disabled = []
+        for k, v in self.__decisions.items():
+            if self.__is_value_enabled(v):
+                enabled.append(k)
+            else:
+                disabled.append(k)
+        return "FeatureFlags(enabled=%r, disabled=%r)" % (enabled, disabled)
```

Comment:
I'd sort the `enabled` and `disabled` lists to ensure that we get a stable - and more importantly readable - `repr`.

Given that you can't actually construct a `FeatureFlags` like that, I'd also have it return `<FeatureFlags: enabled=[...], disabled=[...]>` or similar.

Reply:
> I'd sort the enabled and disabled lists to ensure that we get a stable - and more importantly readable - repr.

It's an ordered dict on Python versions we actually care about, and ordering them in access order seems as good a thing to do as any.

> Given that you can't actually construct a FeatureFlags like that, I'd also have it return <FeatureFlags: enabled=[...], disabled=[...]> or similar.

I went with an alternative approach: Make it so you *can* construct a FeatureFlags object like that. These things are going to end up in our public API anyway eventually and it would be nice to be able to use them with `@example`.